src/sinks/cratedb.rs

```diff
@@ -0,0 +1,556 @@
+use crate::{
+    buffers::Acker,
+    event::{unflatten::MapValue, Event},
+    sinks::util::{
+        http::{HttpRetryLogic, HttpService},
+        retries::FixedRetryPolicy,
+        Batch, BatchServiceSink, SinkExt,
+    },
+    topology::config::{DataType, SinkConfig},
+};
+use futures::{Future, Sink};
+use headers::HeaderMapExt;
+use http::Uri;
+use hyper::{Body, Client, Request};
+use hyper_tls::HttpsConnector;
+use serde::{Deserialize, Serialize, Serializer};
+use std::collections::HashMap;
+use std::time::Duration;
+use string_cache::DefaultAtom as Atom;
+use tower::ServiceBuilder;
+
+#[derive(Deserialize, Serialize, Clone, Debug, Default)]
+#[serde(deny_unknown_fields)]
+pub struct CrateDBSinkConfig {
+    pub uri: String,
+    pub schema: Option<String>,
+    pub table: Option<String>,
+    pub columns: Option<Vec<String>>,
+    pub keys: Option<Vec<String>>,
+    #[serde(flatten)]
+    pub basic_auth: Option<BasicAuth>,
+    pub batch_size: Option<usize>,
+    pub batch_timeout: Option<u64>,
+
+    // Tower Request based configuration
+    pub request_in_flight_limit: Option<usize>,
+    pub request_timeout_secs: Option<u64>,
+    pub request_rate_limit_duration_secs: Option<u64>,
+    pub request_rate_limit_num: Option<u64>,
+    pub request_retry_attempts: Option<usize>,
+    pub request_retry_backoff_secs: Option<u64>,
+}
+
+#[derive(Deserialize, Serialize, Clone, Debug)]
+#[serde(deny_unknown_fields)]
+pub struct BasicAuth {
+    user: String,
+    password: String,
+}
+
+impl BasicAuth {
+    fn apply(&self, header_map: &mut http::header::HeaderMap) {
+        let auth = headers::Authorization::basic(&self.user, &self.password);
+        header_map.typed_insert(auth)
+    }
+}
+
+#[typetag::serde(name = "cratedb")]
+impl SinkConfig for CrateDBSinkConfig {
+    fn build(&self, acker: Acker) -> Result<(super::RouterSink, super::Healthcheck), String> {
+        let sink = cratedb(self.clone(), acker)?;
+
+        let healtcheck = healthcheck(self.uri.clone(), self.basic_auth.clone())?;
+        Ok((sink, healtcheck))
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Log
+    }
+}
+
+#[derive(Debug)]
+enum BufferValue {
+    Statement(String),
+    BulkArgs(Vec<Vec<MapValue>>),
```

Comment:
do you think there would be a way to avoid using MapValue? It was designed as a private internal type that shouldn't be exposed. Curious, if there are any work arounds?

Reply:
I need access to the unflattened event data. Making `MapValue` public was the only way I could find. I'm happy for an alternative.

Reply:
do you think it would be possible to use ValueKind instead?

Reply:
`ValueKind` won't work as it doesn't allow a nested structure. It only supports "[primitive types](https://github.com/timberio/vector/blob/8361f6a36ce604e39ea124b2864060c5cfa680ae/src/event/mod.rs#L169-L175)".

Reply:
Ok that is very fair! I think that is a fine enough use case to expose an internal like that.

cc @lukesteensen let me know if you have any better ideas.

Reply:
It seems likely that we move to a nested internal representation, which I think would essentially merge `MapValue` and `ValueKind`. Given that, I don't think it's a big deal make it public for now.